### PR TITLE
[QoS] Support generating QoS test parameters according to configuration on Mellanox platform

### PR DESCRIPTION
**Pull request by stephenxs**
### Description of PR

Summary:
1. Add a plugin that generates QoS test parameters according to configuration for the Mellanox platform.
   The plugin will be load fixture dutQosConfig
2. Sometimes a deviation can be introduced by calculating test parameters automatically.
   Support passing margin from the qos_params for the purpose of covering the deviation.
3. Support passing packet size from the qos_params for lossy queue and queue shared watermark test.

Signed-off-by: Stephen Sun <stephens@mellanox.com>

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### CODE REVIEW

**Comment by liat-grozovik** [comment]
retest vsimage please

File: tests/qos/files/mellanox/qos_param_generator.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        wm_pg_shared_lossless['pkts_num_trig_pfc'] = pkts_num_dismiss_pfc
+        wm_pg_shared_lossless['cell_size'] = self.cell_size
+
+        wm_q_shared_lossless = self.qos_params_mlnx[self.speed_cable_len]['wm_q_shared_lossless']
+        wm_q_shared_lossless['pkts_num_trig_ingr_drp'] = pkts_num_trig_ingr_drp
+        wm_q_shared_lossless['cell_size'] = self.cell_size
+
+        lossy_queue = self.qos_params_mlnx['lossy_queue_1']
+        lossy_queue['pkts_num_trig_egr_drp'] = pkts_num_trig_egr_drp - 1
+        lossy_queue['cell_size'] = self.cell_size
```
Comment (neethajohn) [inline tests/qos/files/mellanox/qos_param_generator.py:132]: For some of the parameters, I don't see 'self.qos_params_mlnx ' getting updated with the new values

File: tests/qos/test_qos_sai.py [hunk 2]
```diff
@@ -397,6 +414,7 @@ def testQosSaiQSharedWatermark(self, queueProfile, ptfhost, dutTestParams, dutCo
             "pkts_num_leak_out": dutQosConfig["param"][portSpeedCableLength]["pkts_num_leak_out"],
             "pkts_num_fill_min": qosConfig[queueProfile]["pkts_num_fill_min"],
             "pkts_num_trig_drp": triggerDrop,
+            "packet_size": qosConfig[queueProfile]["packet_size"],
```
Comment (neethajohn) [inline tests/qos/test_qos_sai.py:417]: This parameter is not provided for all asics

File: tests/saitests/sai_qos_tests.py [hunk 3]
```diff
@@ -1343,22 +1354,37 @@ def runTest(self):
...
-        default_packet_length = 64
-        pkt = simple_tcp_packet(pktlen=default_packet_length,
+        if 'packet_size' in self.test_params.keys():
+            packet_length = int(self.test_params['packet_size'])
+            cell_size = int(self.test_params['cell_size'])
+            if packet_length != 64:
+                cell_occupancy = (packet_length + cell_size - 1) / cell_size
+                pkts_num_trig_egr_drp /= cell_occupancy
+                # It is possible that pkts_num_trig_egr_drp * cell_occupancy < original pkts_num_trig_egr_drp,
+                # which probaly can fail the assert(xmit_counters[EGRESS_DROP] > xmit_counters_base[EGRESS_DROP])
```
Comment (neethajohn) [inline tests/saitests/sai_qos_tests.py:1364]: spell check 'probably'

File: tests/saitests/sai_qos_tests.py [hunk 4]
```diff
@@ -1654,32 +1689,35 @@ def runTest(self):
...
             send_packet(self, src_port_id, pkt, pkts_num)
             time.sleep(8)
             q_wm_res, pg_shared_wm_res, pg_headroom_wm_res = sai_thrift_read_port_watermarks(self.client, port_list[dst_port_id])
-            print >> sys.stderr, "exceeded pkts num sent: %d, expected watermark: %d, actual value: %d" % (pkts_num, (expected_wm * cell_size), q_wm_res[queue])
-            assert(expected_wm == total_shared)
+            print >> sys.stderr, "exceeded pkts num sent: %d, actual value: %d, expected watermark: %d" % (pkts_num, q_wm_res[queue], ((expected_wm + cell_occupancy) * cell_size))
+            assert(fragment < cell_occupancy)
             assert(expected_wm * cell_size <= q_wm_res[queue])
-            assert(q_wm_res[queue] <= (expected_wm + margin) * cell_size)
+            assert(q_wm_res[queue] <= (expected_wm + margin + cell_occupancy * 2) * cell_size)
```
Comment (neethajohn) [inline tests/saitests/sai_qos_tests.py:1720]: Why is it cell_occupancy * 2?

File: tests/qos/files/mellanox/qos_param_generator.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            'spc3': {
+                'cell_size': 144,
+                'headroom_overhead': 64
+            }
+        }
+
+        self.asic_type = asic_type
+        self.cell_size = asic_param_dic[asic_type]['cell_size']
+        self.headroom_overhead = asic_param_dic[asic_type]['headroom_overhead']
+        if speed_cable_len[0:6] == '400000':
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply tests/qos/files/mellanox/qos_param_generator.py:23]: No. This is for 400G port in MSN4700.

File: tests/qos/files/mellanox/qos_param_generator.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+                ingressLosslessProfile
+                egressLossyProfile
+        """
+        xon = int(math.ceil(float(self.ingressLosslessProfile['xon']) / self.cell_size))
+        xoff = int(math.ceil(float(self.ingressLosslessProfile['xoff']) / self.cell_size))
+        size = int(math.ceil(float(self.ingressLosslessProfile['size']) / self.cell_size))
+        headroom = size
+        hysteresis = headroom - (xon + xoff)
+        ingress_lossless_size = int(math.ceil(float(self.ingressLosslessProfile['static_th']) / self.cell_size)) - headroom
+        egress_lossy_size = int(math.ceil(float(self.egressLossyProfile['static_th']) / self.cell_size))
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply tests/qos/files/mellanox/qos_param_generator.py:59]: This is to take advantage of the current code. In the [qos_sai_base.py:L61](https://github.com/Azure/sonic-mgmt/blob/master/tests/qos/qos_sai_base.py#L61), the `static_th` is updated according to the `dynamic_th`. So `static_th` will always contain the number of bytes in the buffer pool this profile can occupy according to the `dynamic_th`.

File: tests/qos/files/mellanox/qos_param_generator.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    def collect_qos_configurations(self):
+        """
+            Collect qos configuration from the following fixtures
+                ingressLosslessProfile
+                egressLossyProfile
+        """
+        xon = int(math.ceil(float(self.ingressLosslessProfile['xon']) / self.cell_size))
+        xoff = int(math.ceil(float(self.ingressLosslessProfile['xoff']) / self.cell_size))
+        size = int(math.ceil(float(self.ingressLosslessProfile['size']) / self.cell_size))
+        headroom = size
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply tests/qos/files/mellanox/qos_param_generator.py:56]: For Mellanox platform, `size` = `xon` + `xoff` + `hysteresis` (only for 400G port).

File: tests/qos/files/mellanox/qos_param_generator.py (hunk 1 above)
Comment (stephenxs) [reply tests/qos/files/mellanox/qos_param_generator.py:132]: Which one is lost?

File: tests/qos/test_qos_sai.py (hunk 2 above)
Comment (stephenxs) [reply tests/qos/test_qos_sai.py:417]: Do you mean for the non-Mellanox ASICs? Should I add them in this PR?

File: tests/saitests/sai_qos_tests.py (hunk 4 above)
Comment (stephenxs) [reply tests/saitests/sai_qos_tests.py:1720]: `cell_occupancy` represents how many cells for one packet. 
`cell_occupancy * 2` means we tolerance the deviation by 2 packets.

File: tests/qos/files/mellanox/qos_param_generator.py (hunk 1 above)
Comment (neethajohn) [reply tests/qos/files/mellanox/qos_param_generator.py:132]: for eg, wm_pg_shared_lossless, lossy_queue. The values in qos.yml are copied to those dict and updated with new values. How are these getting used in the testcases unless you write them back?

File: tests/qos/test_qos_sai.py (hunk 2 above)
Comment (neethajohn) [reply tests/qos/test_qos_sai.py:417]: yes for non Mellanox asics. use the same logic as in other testcases. update testParams later based on key presence and handle it in the sai_qos_tests.py

File: tests/saitests/sai_qos_tests.py (hunk 4 above)
Comment (neethajohn) [reply tests/saitests/sai_qos_tests.py:1720]: Wouldn't the margin offer some tolerance?

Comment (stephenxs) [reply tests/saitests/sai_qos_tests.py:1720]: Sounds reasonable. Will fixed.

File: tests/qos/files/mellanox/qos_param_generator.py (hunk 1 above)
Comment (stephenxs) [reply tests/qos/files/mellanox/qos_param_generator.py:132]: Yes. The parameters which are not configuration related are copied back from the qos.yml to the dictionary.
The logic in `QosParamMellanox` is like this:
1. Load `qos.yml` into an internal dictionary.
2. Fetch the configuration by using `QosSaiBase`.
3. Generate configuration related parameters and add them to the internal dictionary.
4. The internal dictionary will be returned to `test_qos_sai`, containing both the parameters generated based on configuration and loaded from `qos.yml`.

File: tests/qos/test_qos_sai.py (hunk 2 above)
Comment (stephenxs) [reply tests/qos/test_qos_sai.py]: Fixed.

File: tests/saitests/sai_qos_tests.py [hunk 8]
```diff
@@ -1654,30 +1689,33 @@ def runTest(self):
...
 
                 pkts_num = pkts_inc
 
             # overflow the shared pool
             send_packet(self, src_port_id, pkt, pkts_num)
             time.sleep(8)
             q_wm_res, pg_shared_wm_res, pg_headroom_wm_res = sai_thrift_read_port_watermarks(self.client, port_list[dst_port_id])
-            print >> sys.stderr, "exceeded pkts num sent: %d, expected watermark: %d, actual value: %d" % (pkts_num, (expected_wm * cell_size), q_wm_res[queue])
-            assert(expected_wm == total_shared)
+            print >> sys.stderr, "exceeded pkts num sent: %d, actual value: %d, expected watermark: %d" % (pkts_num, q_wm_res[queue], ((expected_wm + cell_occupancy) * cell_size))
```
Comment (neethajohn) [inline tests/saitests/sai_qos_tests.py:1717]: Can you fix expected watermark value?

File: tests/qos/files/mellanox/qos_param_generator.py (hunk 1 above)
Comment (neethajohn) [reply tests/qos/files/mellanox/qos_param_generator.py:132]: I did see for xon and xoff in lines 99,100,110,111 where after the values were updated they are written back to self.qos_params_mlnx. Don't see that being done for lot of others like wm_pg_shared_lossless, lossy_queue. If we take lossy_queue eg, 'cell size' and 'pkts_num_trig_egr_drp' are updated (lines 130-132) but they are not written back and we return only self.qos_params_mlnx to the testcases

Comment (stephenxs) [reply tests/qos/files/mellanox/qos_param_generator.py:132]: For test cases `xon`, `xoff`, and `wm_shared_lossy`, there are at least 2 group of parameters sharing common parameters. 
Taking `xon` as an example, we have `xon_1` and `xon_2`.
- The common parameters are set into a temporary dict `xon` ([line 102 ~ 105)](https://github.com/Azure/sonic-mgmt/pull/1886/files#diff-59cc3af4e9ac202e45277acd476e10e6R102) and then updated to the original dict `self.qos_params_mlnx['xon_1']` and `self.qos_params_mlnx['xon_2']` by using `update` method. ([line 110, 111](https://github.com/Azure/sonic-mgmt/pull/1886/files#diff-59cc3af4e9ac202e45277acd476e10e6R110)).
- The parameters differing between 2 groups are in the qos.yml and will be loaded into dict in the beginning.

For other test cases that have only 1 group of parameters, all the parameters are set into the original dict directly. For example, `wm_pg_headroom`

File: tests/saitests/sai_qos_tests.py (hunk 8 above)
Comment (stephenxs) [reply tests/saitests/sai_qos_tests.py]: Fixed.

**Pull request merged**